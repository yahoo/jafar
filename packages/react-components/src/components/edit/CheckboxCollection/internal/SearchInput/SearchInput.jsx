import React from 'react';
import PropTypes from 'prop-types';
import Paper from '@material-ui/core/Paper';
import InputBase from '@material-ui/core/InputBase';
import SearchIcon from '@material-ui/icons/Search';
import IconButton from '@material-ui/core/IconButton';
import { withStyles } from '@material-ui/core/styles';
import { noop } from 'lodash';

const styles = {
  root: {
    padding: '2px 4px',
    display: 'flex',
    alignItems: 'center',
  },
  input: {
    marginLeft: 8,
    flex: 1,
  },
  iconButton: {
    padding: 10,
  },
};

class SearchInput extends React.PureComponent {
  static propTypes = {
    value: PropTypes.string,
    placeholder: PropTypes.string,
    disabled: PropTypes.bool,
    onChange: PropTypes.func,
    onSearch: PropTypes.func,
  };

  static defaultProps = {
    value: '',
    placeholder: undefined,
    disabled: false,
    onChange: noop,
    onSearch: noop,
  };

  onChange = (e) => {
    const value = e.target.value;
    this.props.onChange(value);
  };

  onSearchIconClick = () => {
    this.props.onSearch(this.props.value);
  };

  onKeyPress = (e) => {
    if (e.charCode === 13) {
      this.props.onSearch(this.props.value);
    }
  };

  render() {
    const { classes } = this.props;

    return (
      <Paper className={classes.root} elevation={1}>
        <InputBase
          type="search"
          className={classes.input}
          value={this.props.value}
          placeholder={this.props.placeholder}
          disabled={this.props.disabled}
          onChange={this.onChange}
          onKeyPress={this.onKeyPress}
        />
        <IconButton
          className={classes.iconButton}
          disabled={this.props.disabled}
          aria-label="Search"
          onClick={this.onSearchIconClick}
        >
          <SearchIcon />
        </IconButton>
      </Paper>
    );
  }
}

export default withStyles(styles)(SearchInput);
