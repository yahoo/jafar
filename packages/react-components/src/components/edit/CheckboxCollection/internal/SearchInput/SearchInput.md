<h2>This is a sub component of other edit components</h2>
<h4>Simple</h4>

```javascript
const SearchInput = require('SearchInput').default;
initialState = { 
    value: ''
};
<SearchInput 
    value={state.value}
    placeholder="Search"
    onChange={(value) => {
        setState({ value });
        console.log(`onChange: ${value}`);
    }}
    onSearch={(value) => {
        console.log(`onSearch: ${value}`);
    }} />
```

<h4>Disabled</h4>

```javascript
const SearchInput = require('SearchInput').default;
initialState = { 
    value: ''
};
<SearchInput 
    value={state.value} 
    placeholder="Search" 
    disabled={true}
    onSearch={(value) => {
        setState({ value });
        console.log(`onSearch: ${value}`);
    }}/>
```