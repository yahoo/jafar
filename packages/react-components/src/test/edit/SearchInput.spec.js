import React from 'react';
import { mount, shallow } from 'enzyme';
import SearchInput from '../../components/edit/CheckboxCollection/internal/SearchInput';

describe('<SearchInput />', () => {
  let component;
  let onChangeSpy;
  let onSearchSpy;
  const value = 'avocado';
  const placeholder = 'Search Food';
  const disabled = false;

  beforeEach(() => {
    onChangeSpy = jest.fn();
    onSearchSpy = jest.fn();
  });

  it('should render provided data', () => {
    component = shallow(
      getComponent(value, placeholder, disabled, onChangeSpy, onSearchSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('should trigger onChange callback on value change', () => {
    component = mount(
      getComponent(value, placeholder, disabled, onChangeSpy, onSearchSpy)
    );
    const newValue = 'banana';
    component.find('input').simulate('change', { target: { value: newValue } });
    expect(onChangeSpy).toHaveBeenCalledWith(newValue);
  });

  it('should trigger onSearch callback on enter', () => {
    component = mount(
      getComponent(value, placeholder, disabled, onChangeSpy, onSearchSpy)
    );
    component.find('input').simulate('keypress', { charCode: 13 });
    expect(onSearchSpy).toHaveBeenCalledWith(value);
  });

  it('should not trigger onSearch callback on keypress that is not enter', () => {
    component = mount(
      getComponent(value, placeholder, disabled, onChangeSpy, onSearchSpy)
    );
    component.find('input').simulate('keypress', { charCode: 10 });
    expect(onSearchSpy).not.toHaveBeenCalled();
  });

  it('should trigger onSearch callback on search icon click', () => {
    component = mount(
      getComponent(value, placeholder, disabled, onChangeSpy, onSearchSpy)
    );
    component.find('button').simulate('click');
    expect(onSearchSpy).toHaveBeenCalledWith(value);
  });

  function getComponent(value, placeholder, disabled, onChange, onSearch) {
    return (
      <SearchInput
        value={value}
        placeholder={placeholder}
        disabled={disabled}
        onChange={onChange}
        onSearch={onSearch}
      />
    );
  }
});
