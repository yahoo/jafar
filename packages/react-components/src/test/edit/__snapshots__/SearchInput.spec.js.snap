// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> should render provided data 1`] = `
<SearchInput
  classes={
    Object {
      "iconButton": "SearchInput-iconButton-3",
      "input": "SearchInput-input-2",
      "root": "SearchInput-root-1",
    }
  }
  disabled={false}
  onChange={[MockFunction]}
  onSearch={[MockFunction]}
  placeholder="Search Food"
  value="avocado"
/>
`;
